modules/selection/src/main/kotlin/com.r3.corda.lib.tokens.selection/memory/services/VaultWatcherService.kt

```diff
@@ -111,36 +124,131 @@ class VaultWatcherService(private val tokenObserver: TokenObserver,
+                        PageSpecification(numberPages, pageSz)
+                    }
+                    pageSpecifications.add(specification)
+                    return specification
+                }
+                private fun splitDouble(multiple: Double): Pair<Int, Double> {
+                    val intPart = floor(multiple).toInt()
+                    return Pair(intPart, multiple - intPart)
+                }
+                private fun isListInSync(states: List<StateAndRef<FungibleToken>>, specification: PageSpecification): Int {
+                    if (states.isEmpty()) {
+                        return 0
+                    }
+                    if (syncPoints.isEmpty()) {
+                        syncPoints.addFirst(SyncPoint(states.last(), specification, toResultSetIndex(states.lastIndex, specification)))
+                        return 0
+                    }
+                    val highResultSetIndex = toResultSetIndex(states.lastIndex, specification)
+                    val lowResultSetIndex = toResultSetIndex(0, specification)
+
+                    syncPoints.forEach {
+                        val listIndex = toListEntryIndex(it.lastKnownResultSetIndex, specification)
+                        if (listIndex in lowResultSetIndex..highResultSetIndex && states[listIndex] == it.stateAndRef) {
+                            // No change in the list
+                            syncPoints.addFirst(SyncPoint(states.last(), specification, toResultSetIndex(states.lastIndex, specification)))
+                            return listIndex+1 // Return first unread entry index, which is sync pos + 1
+                        }
+                        if (states.contains(it.stateAndRef)) {
+                            // List has changed but we still see the sync point in current list
+                            syncPoints.addFirst(SyncPoint(states.last(), specification, toResultSetIndex(states.lastIndex, specification)))
+                            val pos = states.indexOf(it.stateAndRef)
+                            it.lastKnownResultSetIndex = toResultSetIndex(pos, specification)
+                            return pos + 1 // Return first unread entry index
+                        }
+                    }
+                    LOG.info("Token loading has become out of sync, will re-sync")
+                    return -1
+                }
+                private fun getNewPageSpecification(): PageSpecification {
+                    pageSpecifications.pollLast()
+                    return if (pageSpecifications.isEmpty()) {
+                        syncPoints.clear()
+                        pageSpecifications.add(initialSpecification)
+                        initialSpecification
+                    } else {
+                        pageSpecifications.last
+                    }
+                }
+
+                private fun toResultSetIndex(listEntryIndex: Int, specification: PageSpecification): Int {
```

Comment:
I think we should be using `Long` for the return value here.  i.e. in theory it is `Int * Int` so could exceed the size of `Int`.   Then need to ripple through other data structures.